starter_code/src/movies.js

```diff
@@ -2,14 +2,132 @@
 
 // Iteration 1: Ordering by year - Order by year, ascending (in growing order)
 
+/*const orderByYear = movies.sort(function(a, b) {
+  if (a.year === b.year) {
+    return a.title.localeCompare(b.title);
+  }
+  return a.year - b.year;
+});*/
+function orderByYear(allMovies) {
+  const orderByYear = allMovies.sort(function(a, b) {
+    if (a.year === b.year) {
+      return a.title.localeCompare(b.title);
+    }
+    return a.year - b.year;
+  });
+  if (allMovies.length < 1) {
+    return [];
+  }
+  return orderByYear;
+}
+
+console.log(orderByYear(movies));
+
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct
 
+function howManyMovies(movie) {
+  const howManyMovies = movie
+    .filter(function(movie) {
+      if (
+        movie.director === "Steven Spielberg" &&
+        movie.genre.includes("Drama")
+      ) {
+        return true;
+      }
+    })
+    .map(function(movie) {
+      return movie.title;
+    });
+  return howManyMovies.length;
+}
+
+console.log(howManyMovies(movies));
+
 // Iteration 3: Alphabetic Order - Order by title and print the first 20 titles
 
+function orderAlphabetically(allMovies) {
+  const orderAlphabetically = allMovies
+    .slice()
+    .sort(function(a, b) {
+      return a.title.localeCompare(b.title);
+    })
+    .map(function(movie) {
+      return movie.title;
+    });
+
+  if (allMovies.length < 1) {
+    return [];
+  }
+  if (orderAlphabetically.length > 20) {
+    return orderAlphabetically.slice(0, 20);
+  }
+  return orderAlphabetically;
+}
+
 // Iteration 4: All rates average - Get the average of all rates with 2 decimals
 
+function ratesAverage(allMovies) {
+  if (allMovies.length === 0) {
+    return 0;
+  }
+  const ratesAverage =
+    allMovies.reduce(function(a, b) {
+      if (b.rate) {
+        return b.rate + a;
+      } else {
+        return a + 0;
+      }
+    }, 0) / allMovies.length;
+  return Math.round(ratesAverage * 100) / 100;
+}
+console.log(ratesAverage(movies));
+
 // Iteration 5: Drama movies - Get the average of Drama Movies
 
+function dramaMoviesRate(movie) {
+  const dramaMovies = movie.filter(function(movie) {
+    if (movie.genre.includes("Drama")) {
+      return true;
+    }
+  });
+  return ratesAverage(dramaMovies);
+}
+
 // Iteration 6: Time Format - Turn duration of the movies from hours to minutes
 
 // BONUS Iteration: Best yearly rate average - Best yearly rate average
+function bestYearAvg(movies) {
+  // I order the movies by years.
+  const moviesOrdered = orderByYear(movies);
```

Comment:
Your orderByYear function requires an array that has the property .title to function as it sorts the array by .title. Jasmine feeds your function with this array that doesn't have .title:
[
      { year: 2000, rate: 9 },
      { year: 2000, rate: 8 },
      { year: 1978, rate: 10 },
      { year: 1978, rate: 7 }
    ];

This causes the final mistake in your code that gives the error code: "Can not do .localCompare of undefined". You'd have to work around this to get it perfect:)